ec2/instance.py

```diff
@@ -14,165 +14,203 @@
 
 
 @operation
-def stop(**kwargs):
+def stop(retry_interval, **_):
     """ Stops an existing EC2 instance.
         If already stopped, this does nothing.
+        Requires:
+            ctx.instance.runtime_properties['aws_resource_id']
     """
     ec2_client = connection.EC2ConnectionClient().client()
 
-    instance_id = ctx.instance.runtime_properties['instance_id']
+    instance_id = ctx.instance.runtime_properties.get('aws_resource_id')
 
     ctx.logger.info('Stopping EC2 Instance.')
     ctx.logger.debug('Attempting to stop EC2 Instance.'
                      '(Instance id: {0}.)'.format(instance_id))
 
     try:
-        instances = ec2_client.stop_instances(instance_id)
-    except (EC2ResponseError, BotoServerError) as e:
+        ec2_client.stop_instances(instance_id)
+    except (boto.exception.EC2ResponseError,
+            boto.exception.BotoServerError) as e:
         raise NonRecoverableError('Error. Failed to stop EC2 Instance: '
-                                  'API returned: {0}.'.format(e))
-
-    if utils.validate_instance_id(instance_id, ctx=ctx):
-        utils.validate_state(instances[0], INSTANCE_STOPPED,
-                             STOP_TIMEOUT, CHECK_INTERVAL, ctx=ctx)
+                                  'API returned: {0}.'.format(str(e)))
+    finally:
+        ctx.instance.runtime_properties.pop('private_dns_name')
+        ctx.instance.runtime_properties.pop('public_dns_name')
+        ctx.instance.runtime_properties.pop('public_ip_address')
+        ctx.instance.runtime_properties.pop('ip')
+
+    # get the timeout code validate state
+    if utils.get_instance_state(ctx=ctx) == 80:
+        ctx.logger.info('Instance {0} is stopped.'.format(instance_id))
+    else:
+        raise RecoverableError('Waiting for server to stop'
+                               'Retrying...',
+                               retry_after=retry_interval)
 
 
 @operation
-def terminate(**kwargs):
+def terminate(retry_interval, **_):
     """ Terminates an existing EC2 instance.
         If already terminated, this does nothing.
     """
     ec2_client = connection.EC2ConnectionClient().client()
 
-    instance_id = ctx.instance.runtime_properties['instance_id']
+    instance_id = ctx.instance.runtime_properties.get('aws_resource_id')
 
     ctx.logger.info('Terminating EC2 Instance.')
     ctx.logger.debug('Attempting to terminate EC2 Instance.'
                      '(Instance id: {0}.)'.format(instance_id))
 
     try:
-        instances = ec2_client.terminate_instances(instance_id)
-    except (EC2ResponseError, BotoServerError) as e:
+        ec2_client.terminate_instances(instance_id)
+    except (boto.exception.EC2ResponseError,
+            boto.exception.BotoServerError) as e:
         raise NonRecoverableError('Error. Failed to terminate '
                                   'EC2 Instance: API returned: {0}.'
-                                  .format(e))
+                                  .format(str(e)))
+    finally:
```

Comment:
only `aws_resource_id` is relevant here. Also, shouldn't be in finally but rather as a consequence of the validation.